### ISSUE TITLE
📘 What would you like to learn about React Native?

### ISSUE THREAD

**Issue by cpojer** [facebook/react-native-website#917]
We are planning on revamping our guides for React Native: https://facebook.github.io/react-native/docs/getting-started

Many of the existing guides and the overall flow of the documentation were written in 2015 and do not reflect well what it is like to build apps with React Native in 2019. We have an idea of what we'd like to improve and will share a plan soon but we'd like to hear from the community what they'd like to see in the React Native documentation:

If you have recently learned how to use React Native or if you imagine learning React Native from scratch and would like to help, here are some questions to get started:
* Which parts that you found out were important were missing from the website?
* Which things would you remove from the website because they are outdated and confusing?
* Which pages could use more detail or a rewrite?
* Which docs were really confusing and hard to understand?
* What are the most helpful resources you used to learn React Native?
* How would you change the guides section on the website? What is the ideal flow to learn React Native? Which things do you think should be taught in which order?

Feel free to reply here with your suggestions and ideas! Thank you 🙏

**Comment by Divine1** [comment]
i'm not able to find docs on the android/ios sdk versions & nodejs,npm versions supported by each expo releases. 

i'm starting to think below issue is because of expo-cli not being compatible with node 10.15.3, because the expo-cli installed without any issue in node 8.11.* . i didnot check in node 9.*. 

https://github.com/expo/expo-cli/issues/412


if there is any docs available for this then please point me to it

**Comment by yousefelgoharyx** [comment]
mmmm, ig better UI would be great

**Comment by ecreeth** [comment]
## What I would like more is to improve the documentation, more components and a proper implementation of the navigation.
#### What parts that you discovered that were important were missing from the website?
- A better design
- Detail a little more the operation of the APIs and components

#### What things would you remove from the website because they are outdated and confusing?
I consider that the OBSOLETE components should be eliminated
- [ListView](https://facebook.github.io/react-native/docs/listview)
- [MaskedViewIOS](https://facebook.github.io/react-native/docs/maskedviewios)
- [BackAndroid](https://facebook.github.io/react-native/docs/backandroid)
- [SnapshotViewIOS](https://facebook.github.io/react-native/docs/snapshotviewios)
- [IimageStore](https://facebook.github.io/react-native/docs/imagestore)
- [VibrationIOS](https://facebook.github.io/react-native/docs/vibrationios)

 #### Which pages could use more details or rewrite?
 - [Animated](https://facebook.github.io/react-native/docs/animated)
 - [Gesture Responder System](https://facebook.github.io/react-native/docs/gesture-responder-system)
- [Building From Source](https://facebook.github.io/react-native/docs/building-from-source)
- [StatusBarIOS](https://facebook.github.io/react-native/docs/statusbarios)
- [Settings](https://facebook.github.io/react-native/docs/settings)
- [Easing](https://facebook.github.io/react-native/docs/easing)
- [Stylesheet](https://facebook.github.io/react-native/docs/stylesheet)
- [Share](https://facebook.github.io/react-native/docs/share)
- [LayoutAnimation](https://facebook.github.io/react-native/docs/layoutanimation)
- [Systrace](https://facebook.github.io/react-native/docs/systrace)
- [Vibration](https://facebook.github.io/react-native/docs/vibration)
- [VirtualizedList](https://facebook.github.io/react-native/docs/virtualizedlist)
- [AppRegistry](https://facebook.github.io/react-native/docs/appregistry)
- [Linking](https://facebook.github.io/react-native/docs/linking)
- [PushNotificationIOS](https://facebook.github.io/react-native/docs/pushnotificationios)
- [panresponder](https://facebook.github.io/react-native/docs/panresponder)
- [interactionmanager](https://facebook.github.io/react-native/docs/interactionmanager)
- [pickerios](https://facebook.github.io/react-native/docs/pickerios)
- [ProgressViewIOS](https://facebook.github.io/react-native/docs/progressviewios)
- [ImagePickerIOS](https://facebook.github.io/react-native/docs/imagepickerios)
- [Statusbar](https://facebook.github.io/react-native/docs/statusbar)
- [ImageEditor](https://facebook.github.io/react-native/docs/imageeditor)
- [Geolocation](https://facebook.github.io/react-native/docs/geolocation)
- [Backhandler](https://facebook.github.io/react-native/docs/backhandler)
- [DatePickerAndroid](https://facebook.github.io/react-native/docs/datepickerandroid)
- [Dimensions](https://facebook.github.io/react-native/docs/dimensions)
- [Switch](https://facebook.github.io/react-native/docs/switch)
- [ViewPagerAndroid](https://facebook.github.io/react-native/docs/viewpagerandroid)
- [Clipboard](https://facebook.github.io/react-native/docs/clipboard)
- [AccessibilityInfo](https://facebook.github.io/react-native/docs/accessibilityinfo)
- [ActionsheetIOS](https://facebook.github.io/react-native/docs/actionsheetios)

#### What are the most useful resources you used to learn React Native?
- Tutorials
- Books
Because the documentation does not offer much, most of the time to see tutorials and read books.

 #### What is the ideal flow to learn React Native?
I think that if the components and APIs are well documented, everyone would find the way they feel most comfortable.

**Comment by orta** [comment]
I learned by diving through the f8 app via https://makeitopen.com - I think a video series where you guide someone through building a small(er) and self contained app, wrapping up with showing how to write some tests and feel comfortable looking at it in iOS/Android sims. 

Something like this probably already exists, but could be re-done with some more production values and be classed as official

**Comment by tsdmrfth** [comment]
When i new to React Native,  as i remember i was very stuggling when i try to make some components that's why i would like to see working examples in both android and iOS with the example codes.

**Comment by Steffi3rd** [comment]
> ## What I would like more is to improve the documentation, more components and a proper implementation of the navigation.

#### What parts that you discovered that were important were missing from the website?
* Storybook like
* Demo with Expo is very bad UX (waiting... ui... interaction...)
* Best community packages
* Examples like : https://reactnativeexample.com/

#### What things would you remove from the website because they are outdated and confusing?
* **CRNA**  : For a new React Native developer, it was VERY interesting at first. But when you need to interact with "Native features" of the device (like Camera, GPS etc.), it's very confusing. I would suggest to not make a difference at the beginning. One way, and decide to `link` something if needs. That's mean : No ejection stuff.

#### Which pages could use more details or rewrite?
* [Animated](https://facebook.github.io/react-native/docs/animated)
* [Stylesheet](https://facebook.github.io/react-native/docs/stylesheet)
* [Navigation](https://facebook.github.io/react-native/docs/stylesheet)

Internal React Native solutions became Community solutions still must have a full detail documentation (Like `react-navigation`).

Some iOS 12 features like Large Header should be available right in this documentation.

#### What are the most useful resources you used to learn React Native?
* The greatest : [Reactiflux](https://www.reactiflux.com)
* React Native Documentation
* Tweets about React

#### What is the ideal flow to learn React Native?
I was easier for me to learn React Native because I already new ReactJS.
And I was waiting for Hooks before starting React Native. 

Anyway, thanks a lot for all of this guys! Keep going, it's amazing!

**Comment by saxenanickk** [comment]
There has to be some good documentation on ART.

**Comment by pie6k** [comment]
I've 'wasted' literally 1000's of lines of code because I didn't know about `LayoutAnimation` and I was creating complex animations (like reordering, unmounting) myself. I know it's my fault but it's so useful that it could be more exposed (as ex-web dev I didn't expect such things to be possible).

Maybe something like 'if you're web dev - this is MUST READ for you'

**Comment by glauberfc** [comment]
> Which parts that you found out were important were missing from the website?

1. You could add more details on *Contributing* docs section, teaching people how use RNTester.
2. I think that will be great if you teach people how Google Play and App Store works, and how deploy an app these stores.
3. It would be nice if there is a section with a tutorial like [makeitopen](https://makeitopen.com/).

**Comment by ofiron01** [comment]
Best practices for debugging, HOW-TOs, known issues, troubleshooting

**Comment by ghmeec** [comment]
Adding App testing in real devices such as via expo, would be great

**Comment by rrramanan** [comment]
• Important that missing from website? 
- Detailed page that can help newbies about important areas (eg. nav, redux, RNBase) info listed and let user redirects from there.  
- Left pane (menu) UI improvement 
- Dark mode

**Comment by AndreiCalazans** [comment]
**Which parts that you found out were important were missing from the website?**

 - Cover the essential knowledge one is expected to have when dealing with Xcode and Android Studio for React Native Apps. Many Devs coming from Web get completely lost in these environments.

- How to add fonts.
- How to set up Launch/Splash screen.
- How to set up Launch Icons.
- For the FlatList component would be nice to have more information on how to handle an ever loading list or infinite list. 

**Which docs were really confusing and hard to understand?**

- Timermixin https://facebook.github.io/react-native/docs/timers#timermixin
- KeyboardAvoidingView (This has always been confusing, a lot of try and error necessary here) https://facebook.github.io/react-native/docs/keyboardavoidingview

- Components which have their native only counterparts. e.g. Picker and PickerIOS.


**What are the most helpful resources you used to learn React Native?**
Tutorials. Currently inside the Basics sections we have small tutorials covering separate parts, beyond this, It would add value to the documentation if some tutorials covering more advanced scenarios were added, e.g. Building a complete layout, Making a news feed with details page, Debugging a native issue, Adding a native module to your App, Make a clone of something which uses different APIs. (I know some of these tutorials are covered by the community, however, most beginners have a hard time finding these resources we could add links to them in the more resources page.)

**How would you change the guides section on the website? What is the ideal flow to learn React Native? Which things do you think should be taught in which order?**

```
**The Basics**
* Getting Started
* Learn the Basics
* Getting serious  // In here we could build something more real world instead of a Hello World App. 
* Props
* State
* Style
* Height and Width
...
```

**Comment by amielmon** [comment]
More documentation on event listeners!! More specifically scrolling actions.

**Comment by Andarius** [comment]
More documentation on flexbox like the one [here](https://medium.com/wix-engineering/the-full-react-native-layout-cheat-sheet-a4147802405c)

**Comment by ecreeth** [comment]
> More documentation on flexbox like the one [here](https://medium.com/wix-engineering/the-full-react-native-layout-cheat-sheet-a4147802405c).
> 
> Some documentation on testing could be nice (with a simple example using Jest and optionally other libs like Enzyme).
> 
> Some links to get started with RN using Typescript (like [here)](https://github.com/Microsoft/TypeScript-React-Native-Starter)

**This is the best place for learn Flexbox**
https://yogalayout.com/

**Comment by Andarius** [comment]
> > More documentation on flexbox like the one [here](https://medium.com/wix-engineering/the-full-react-native-layout-cheat-sheet-a4147802405c).
> 
> **This is the best place for learn Flexbox**
> https://yogalayout.com/

I prefer to have a quick overview like [![flexbox.png](https://i.postimg.cc/c4JvFTdn/flexbox.png)](https://postimg.cc/7bpP63GH)

**Comment by e-molochko** [comment]
Please, add more detailed description for component's/method's properties and their descriptions

**Comment by CodingItWrong** [comment]
I'd love a page with quicker access to the docs for each component. Usually when i go to the docs I want to look up a specific component, but I have to scroll down in the sidebar to get to the Components section. If there was a Components or API page that just had those Components as the only things in the sidebar, that'd make navigation quicker. Kind of like http://api.emberjs.com/ember/release

**Comment by sunday58** [comment]
I did love to see more images on the documentation, cause image explains practically than just text..

And more explained documentation on intent to make your app access other default apps on your phone..example; Camera, gallery, musicplayer etc.

**Comment by tiancihe** [comment]
> I learned by diving through the f8 app via https://makeitopen.com - I think a video series where you guide someone through building a small(er) and self contained app, wrapping up with showing how to write some tests and feel comfortable looking at it in iOS/Android sims.
> 
> Something like this probably already exists, but could be re-done with some more production values and be classed as official

IMO, detailed docs are way more better than videos. So, my suggestion is to focus the effort on the doc rather than videos, and maybe a whole lot more detail on how to set up properly, maybe even how to set up for react-native-windows.

**Comment by dev6james** [comment]
This comes to mind immediately (LayoutAnimation) https://github.com/facebook/react-native-website/issues/889

Beyond that,  more on the native debbuging and profiling experience.  This tiny amount in the docs makes a new RN dev feel hopeless the second they run into a tricky native issue.  Even pointing to guides in the android and iOS docs would be handy if anything more is considered out of scope.  But honestly I think small guides on basic Xcode and Android Studio debugging would be worthwhile.

**Comment by tijs** [comment]
I would love to see a guide on packaging up react native code as re-usable NPM packages. There are some short references to create-library but no step by step reference. Ideally this would cover:

* packaging up pure JS code
* adding (natively linked) dependencies
* packaging up native modules 
* adding flow and/or typescript definitions

**Comment by ramyareye** [comment]
First as a web developer I think I'm happy that I started with CRNA instead of expo! Why? cause it forced me to dive in and learn faster, I don't know why but still I can't find reason to go with Expo. I think working on react-native requires knowledge about native codes.

Second I think there should be a complete guide to start with Visual Studio Code, it saves bunch of time.

Third is about learning about layouts, what better than https://flexboxfroggy.com/ to suggest ?

Fourth is about React and React Native. I saw people started directly with RN with no React Knowledge. So there should be a guide to have a fast course on React and to see what's going on! I saw people who were trying to find some guid for Redux on RN! I should there should be smart course for beginners like this.

Regards

**Comment by Fausto95** [comment]
Well documented native modules in react-native with Swift and Kotlin

**Comment by RoyEde** [comment]
My suggestion is to add more examples whenever possible, at least for common usage.

* Things like `LayoutAnimation` can be absolutely confusing without proper examples.

* Some common patterns would be appreciated (for example, with the `Animated` API).

We absolutely need a scroll back to top button on the page. Or at least a button that goes straight to the components props index again.

**Comment by michalchudziak** [comment]
I think a bit deeper documentation of native classes of React Native might be helpful, especially when it comes to integration with native apps. I've had to figure out a lot by scouring through the codebase, perhaps with well-documented classes, the entry level would decrease :)

**Comment by pavermakov** [comment]
It would be nice to have more info about how the AppState works  in react-native. For example, how it behaves when native popups show up on android and the AppState changes all of a sudden. I have no prior experience with java, so I was frustrated when I first encountered this.

Also, it would be nice to include screenshots for all available components.

**Comment by cpojer** [comment]
Thanks everyone for all of your suggestions. I now created an umbrella task to track the website rewrite: https://github.com/facebook/react-native-website/issues/929

We need a **ton** of help to make this happen. If you are interested, check out the umbrella issue and send some PRs :)
